runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -410,14 +416,18 @@ protected WorkItemCommitRequest persistDirectly(WindmillStateCache.ForKey cache)
         return WorkItemCommitRequest.newBuilder().buildPartial();
       }
 
-      ByteString.Output stream = ByteString.newOutput();
-      if (value != null) {
-        coder.encode(value, stream, Coder.Context.OUTER);
+      ByteString encoded = null;
+      if (cachedSize == -1 || modified) {
```

Comment:
This logic is really frustrating that we are encoding because we aren't returning the size of what we read as part of the state future from WindmillStateReader.

This could be better but this is definitely an improvement to not re-encode the value needlessly.

Reply:
agreed, it took me awhile to reason this out w/ the different code paths that can affect the values here.  I'm reasonably sure you don't need the `cachedSize == -1` branch but, it's been working like this for ~1 year so I didn't want to tempt fate.

There's also the case where this is the first write, but `valueIsKnown = false` (because `isNewKey = false`), so we end up with a read that doesn't return anything.

Thanks for the quick reviews!